File: Data Structures & Algorithms/two-integer-sum-ii/submission-0.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        n = 0
        j = len(nums)-1
        while n < len(nums):
            if nums[n]+nums[j]==target:
                return [n+1,j+1]
            if nums[n]+nums[j]>target:
                 j = j-1
            else:
                 n = n +1

File: Data Structures & Algorithms/target-sum/submission-0.java
class Solution {
    public int ways(int[] nums, int target, int index, int[][] dp, int offset) {
        if (index == nums.length) {
            return target == 0 ? 1 : 0;
        }
        if (target > offset || target < -offset) {
            return 0;
    }
        if (dp[index][target + offset] != -1) {
            return dp[index][target + offset];
        }

        int add = ways(nums, target - nums[index], index + 1, dp, offset);
        int subtract = ways(nums, target + nums[index], index + 1, dp, offset);

        return dp[index][target + offset] = add + subtract;
    }

    public int findTargetSumWays(int[] nums, int target) {
        int totalSum = 0;
        for (int num : nums) totalSum += num;

        if (Math.abs(target) > totalSum) return 0;

        int[][] dp = new int[nums.length][2 * totalSum + 1];
        for (int[] row : dp) Arrays.fill(row, -1);

        return ways(nums, target, 0, dp, totalSum);
    }
}